include/dbgroup/atomic/mwcas/lock_free/mwcas_descriptor.hpp

```diff
@@ -0,0 +1,252 @@
+   * the MwCAS function.
+   */
+  [[nodiscard]] static auto GetDescriptor()  //
+      -> MwCASDescriptor *;
+
+  /*############################################################################
+   * Public utility functions
+   *##########################################################################*/
+
+  /**
+   * @brief Read a value from a given memory address.
+   *
+   * @tparam T An expected class of a target field.
+   * @param addr A target memory address to read.
+   * @param fence A flag for controling std::memory_order.
+   * @return A read value.
+   * @note If a memory address is included in MwCAS target fields, it must be
+   * read via this function.
+   */
+  template <class T>
+  static auto
+  Read(  //
+      const void *addr,
+      const std::memory_order fence = std::memory_order_seq_cst)  //
+      -> T
+  {
+    static_assert(CanMwCAS<T>());
+
+    const auto *target_addr = static_cast<const std::atomic_uint64_t *>(addr);
+    uint64_t word{};
+    while (true) {
+      for (size_t i = 1; true; ++i) {
+        word = target_addr->load(fence);
+        if ((word & kMwCASFlag) == 0) return std::bit_cast<T>(word);
+        if (i > kRetryNum) break;
+        CPP_UTILITY_SPINLOCK_HINT
+      }
+      std::this_thread::sleep_for(kBackOffTime);
+    }
+  }
+
+  /**
+   * @brief Add a new MwCAS target to this descriptor.
+   *
+   * @tparam T The class of a target word.
+   * @param addr A target memory address.
+   * @param old_val The expected value of a target field.
+   * @param new_val An inserting value into a target field.
+   * @param fence A flag for controling std::memory_order.
+   */
+  template <class T>
+  constexpr void
+  AddMwCASTarget(  //
+      void *addr,
+      const T old_val,
+      const T new_val,
+      const std::memory_order fence = std::memory_order_seq_cst)
+  {
+    static_assert(CanMwCAS<T>());
+
+    targets_.at(target_cnt_++) =
+        MwCASTarget{static_cast<std::atomic_uint64_t *>(addr), old_val, new_val, fence};
+  }
+
+  /**
+   * @brief Perform a MwCAS operation by using registered targets.
+   *
+   * @retval true if a MwCAS operation succeeds.
+   * @retval false otherwise.
+   */
+  auto MwCAS()  //
+      -> bool;
+
+ private:
+  /*############################################################################
+   * Internal types
+   *##########################################################################*/
+
+  /**
+   * @brief An enumeration for representing MwCAS status.
+   *
+   */
+  enum Status : uint64_t {
+    kUndecided = 0,
+    kSucceeded,
+    kFailed,
+  };
+
+  /**
+   * @brief A class for representing MwCAS targets.
+   *
+   */
+  struct MwCASTarget {
+    /// @brief A target memory address.
+    std::atomic_uint64_t *addr;
+
+    /// @brief An expected value of a target field.
+    uint64_t old_val;
+
+    /// @brief An inserting value into a target field.
+    uint64_t new_val;
+
+    /// @brief access count for GC.
+    uint64_t cnt;
```

Comment:
`MwCASTarget` は全体で32バイトに納めたいので，4バイトで管理しましょう．また，後々atomicに操作する必要もでてきます．

```suggestion
    std::atomic_uint32_t cnt;
```